// @keystonejs/fields/types/Text/views/Field/dist/fields.cjs.prod.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: !0
});

var core = require("@emotion/core"), fields = require("@arch-ui/fields"), input = require("@arch-ui/input");

const TextField = ({onChange: onChange, autoFocus: autoFocus, field: field, errors: errors, value: value = "", isDisabled: isDisabled}) => {
  const {isMultiline: isMultiline} = field.config, htmlID = "ks-input-" + field.path, canRead = errors.every(error => !(error instanceof Error && "AccessDeniedError" === error.name)), error = errors.find(error => error instanceof Error && "AccessDeniedError" === error.name);
  return core.jsx(fields.FieldContainer, null, core.jsx(fields.FieldLabel, {
    htmlFor: htmlID,
    field: field,
    errors: errors
  }), core.jsx(fields.FieldDescription, {
    text: field.adminDoc
  }), core.jsx(fields.FieldInput, null, core.jsx(input.Input, {
    autoComplete: "off",
    autoFocus: autoFocus,
    required: field.isRequired,
    type: "text",
    value: canRead ? value : void 0,
    placeholder: canRead ? void 0 : error.message,
    onChange: event => {
      onChange(event.target.value);
    },
    id: htmlID,
    isMultiline: isMultiline,
    disabled: isDisabled
  })));
};

exports.default = TextField;
