// @keystonejs/fields/types/Checkbox/views/Field/dist/fields.cjs.prod.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: !0
});

var core = require("@emotion/core"), fields = require("@arch-ui/fields"), controls = require("@arch-ui/controls");

const CheckboxField = ({onChange: onChange, autoFocus: autoFocus, field: field, value: value, errors: errors, isDisabled: isDisabled}) => {
  const checked = value || !1, htmlID = "ks-input-" + field.path;
  return core.jsx(fields.FieldContainer, null, core.jsx(fields.FieldDescription, {
    text: field.adminDoc
  }), core.jsx(fields.FieldInput, {
    css: {
      height: 35,
      alignItems: "center"
    }
  }, core.jsx(controls.CheckboxPrimitive, {
    autoFocus: autoFocus,
    checked: checked,
    onChange: event => {
      onChange(event.target.checked);
    },
    id: htmlID,
    isDisabled: isDisabled
  }), core.jsx(fields.FieldLabel, {
    htmlFor: htmlID,
    field: field,
    errors: errors,
    css: {
      padding: "4px",
      fontSize: "1rem",
      width: "100%"
    }
  })));
};

exports.default = CheckboxField;
