// @keystonejs/fields/types/Float/views/Field/dist/fields.cjs.dev.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var fields = require('@arch-ui/fields');
var React = require('react');
var React__default = _interopDefault(React);
var input = require('@arch-ui/input');

const FloatField = ({
  onChange,
  autoFocus,
  field,
  value,
  errors,
  isDisabled
}) => {
  const handleChange = event => {
    const value = event.target.value; // Similar implementation as per old Keystone version

    if (/^-?\d*\.?\d*$/.test(value)) {
      onChange(value);
    }
  };

  const valueToString = value => {
    // Make the value a string to keep react happy.
    if (typeof value === 'string') {
      return value;
    } else if (typeof value === 'number') {
      return String(value);
    } else {
      // If it is neither string nor number then it must be empty.
      return '';
    }
  };

  const htmlID = `ks-input-${field.path}`;
  return React__default.createElement(fields.FieldContainer, null, React__default.createElement(fields.FieldLabel, {
    htmlFor: htmlID,
    field: field,
    errors: errors
  }), React__default.createElement(fields.FieldDescription, {
    text: field.adminDoc
  }), React__default.createElement(fields.FieldInput, null, React__default.createElement(input.Input, {
    autoComplete: "off",
    autoFocus: autoFocus,
    required: field.isRequired,
    type: "text",
    value: valueToString(value),
    onChange: handleChange,
    id: htmlID,
    disabled: isDisabled
  })));
};

exports.default = FloatField;
