// @keystonejs/fields/types/Decimal/views/Field/dist/fields.cjs.prod.js
"use strict";

function _interopDefault(ex) {
  return ex && "object" == typeof ex && "default" in ex ? ex.default : ex;
}

Object.defineProperty(exports, "__esModule", {
  value: !0
});

var fields = require("@arch-ui/fields"), input = require("@arch-ui/input"), React = require("react"), React__default = _interopDefault(React);

const DecimalField = ({onChange: onChange, autoFocus: autoFocus, field: field, value: value, errors: errors, isDisabled: isDisabled}) => {
  const {symbol: symbol} = field.config, htmlID = "ks-input-" + field.path;
  return React__default.createElement(fields.FieldContainer, null, React__default.createElement(fields.FieldLabel, {
    htmlFor: htmlID,
    field: field,
    errors: errors
  }), React__default.createElement(fields.FieldDescription, {
    text: field.adminDoc
  }), React__default.createElement(fields.FieldInput, null, symbol && React__default.createElement(fields.Currency, null, symbol), React__default.createElement(input.Input, {
    autoComplete: "off",
    autoFocus: autoFocus,
    required: field.isRequired,
    type: "text",
    value: (value => "string" == typeof value ? value : "number" == typeof value ? String(value) : "")(value),
    onChange: event => {
      const value = event.target.value;
      onChange(value.replace(/[^0-9.,]+/g, ""));
    },
    id: htmlID,
    disabled: isDisabled
  })));
};

exports.default = DecimalField;
