// @keystonejs/fields/types/Relationship/views/Filter/dist/fields.cjs.prod.js
"use strict";

function _interopDefault(ex) {
  return ex && "object" == typeof ex && "default" in ex ? ex.default : ex;
}

Object.defineProperty(exports, "__esModule", {
  value: !0
}), require("@babel/runtime/helpers/defineProperty"), require("@babel/runtime/helpers/objectWithoutProperties"), 
require("@emotion/core");

var React = require("react"), React__default = _interopDefault(React);

require("@babel/runtime/helpers/extends"), require("@apollo/client"), require("@arch-ui/select"), 
require("react-select"), require("intersection-observer");

var RelationshipSelect = require("../../../../../dist/RelationshipSelect-a7bf5d1a.cjs.prod.js");

const EventCatcher = ({children: children}) => React__default.createElement("div", {
  onClick: e => {
    e.preventDefault(), e.stopPropagation();
  }
}, children), RelationshipFilterView = ({onChange: onChange, filter: filter, field: field, value: value}) => {
  if (!filter) return null;
  const htmlID = "ks-input-" + field.path;
  return React__default.createElement(EventCatcher, null, React__default.createElement(RelationshipSelect.RelationshipSelect, {
    field: field,
    renderContext: null,
    htmlID: htmlID,
    onChange: option => {
      if (null === option) onChange(null); else {
        const {value: value} = option;
        value && onChange(value.id);
      }
    },
    value: value,
    isMulti: !1
  }));
};

exports.default = RelationshipFilterView;
